### ISSUE TITLE
Remove the need for relationship repositories

### ISSUE THREAD

**Issue by chb0github** [katharsis-project/katharsis-framework#121]
The title says it all. I have brought this up and it's been discussed, but I am making it an official issue so we can continue to conversation and throw some code around on how to do it.

Once you have a `Repository<Something,Long>` you don't need anything else. And I am pretty sure the `Long` can go to.

**Comment by remmeier** [comment]
I don't think there is a general and easy answer to that. It depends on the use case. For many, in particular simpler, use cases the relationship repositories might not necessarly required. For others they are fairly vital. It can have its advantages to have data and relationships separate. At some point the json api might standardize a bulk update operation in some form or other. There it can play a vital role to sort out the order how updates are applied. Or for relationship with a skewed distribution, like a many-to-many relationship where a single resource is related to millions of others. Many-to-many relations are a problem in general without relationship repositories. The current design also closely resembles the json api spec, which I consider a good thing.

So relationship repositories should be supported like it is the case now from my perspective. But it should be possible to have "implict" relationship repositories that handle everything trough the respective resource repositories. The implict repositories could be used if no explicit one are defined. This would provide most flexibility.

I'm not particuly keen about the Long generic type argument either.

**Comment by mohlek** [comment]
The "Long" generic type is in my case in heavy use.
Some entities have String id's and some have Integer id's and some have Long id's.

**Comment by masterspambot** [comment]
I think @remmeier has good point here. And if combine both of those approaches we will find ultimate solution. So having out of the box relationship repositories with possibility to override the generic implementation seems perfect to me. :+1:

**Comment by chb0github** [comment]
The type of the `@JsonApiId` can be determined at runtime. So, I wonder how necessary it is. 

I look at the implementation of my `RelationshipRepository`ies and they are no more than 3 lines of code. All similar. In addition, I see what spring-data-rest can do and that is nearly zenlike. Each repository is an 

```
@RepositoryRestResource(collectionResourceDescription = @Description("Organization Description"))
public interface OrganizationRepository extends PagingAndSortingRepository<Organization, Long> {}
```
And from there I get an entire crud. That is ideal. 

But, an "advanced" use case support is wise. So, I guess what Is needed is this hall-of-mirrors implementation. 

Honestly, this really gets to how we discover repositories. Because it would be during that processes that you could/would apply such awesome-sauce

**Comment by remmeier** [comment]
I was just working in the area, so I also took some steps in the discussed direction.

I implemented the new abstract base classes:

- QuerySpecResourceRepositoryBase
- QuerySpecRelationshipRepositoryBase

For a fully-functional repository, just one findAll method for QuerySpecResourceRepositoryBase needs to be implemented. All the other finders are handled with querySpec. And QuerySpecRelationshipRepositoryBase make use of PropertiesUtils and the resource repositories on both sides to implement its functionality (assuming that the resource repositories save the relationships as well).

One further minor change was the introduction of "opposite" for JsonApiToMany and JsonApiToOne so that QuerySpecRelationshipRepositoryBase  can do its magic.

=> An example is available with ProjectQuerySpecRepository. It starts to get really simple to create a repository,,, (https://github.com/katharsis-project/katharsis-framework/pull/200)

A a next step, QuerySpecRelationshipRepositoryBase  could be instantiated dynamically if no custom implementation was provided.
